tools/test-setup.sh

```diff
@@ -0,0 +1,31 @@
+#!/bin/bash
+
+# This tool is used to setup the environment for running the tests. Its name
+# name and location is based on Zuul CI, which can automatically run it.
+set -euo pipefail
+
+# User specific environment
+# shellcheck disable=SC2076
+if ! [[ "$PATH" =~ "$HOME/.local/bin" ]]; then
+    PATH="$HOME/.local/bin:$PATH"
+fi
+
+if [ -f "/usr/bin/apt-get" ]; then
+    if [ ! -f "/var/cache/apt/pkgcache.bin" ]; then
+        sudo apt-get update # mandatory or other apt-get commands fail
+    fi
+    # avoid outdated ansible and pipx
+    sudo apt-get remove -y ansible pipx || true
+    # cspell:disable-next-line
+    sudo apt-get install -y --no-install-recommends -o=Dpkg::Use-Pty=0 \
+        git python3-venv python3-pip
+fi
+
+which pipx || python3 -m pip install --user pipx
+which -a pipx
+which tox || pipx install tox
```

Comment:
I guess here is a convention thingy, why use `pipx` over the regular `pip` version.

Second comment is that why are we installing the software that i would assume that github-ci would provide for us? is pip not included by default in their python installation that follows along with the image we run?

Third is that why do we need to run it inside a separate venv/pipx isolation when we already run it in isolation inside github-ci itself? why the extra layer i mean